samza-core/src/main/scala/org/apache/samza/job/local/ProcessJob.scala

```diff
@@ -19,113 +19,118 @@
 
-import org.apache.samza.SamzaException
 import org.apache.samza.coordinator.JobModelManager
-import org.apache.samza.job.ApplicationStatus.{New, Running, UnsuccessfulFinish}
+import org.apache.samza.job.ApplicationStatus.{New, Running, SuccessfulFinish, UnsuccessfulFinish}
 import org.apache.samza.job.{ApplicationStatus, CommandBuilder, StreamJob}
 import org.apache.samza.util.Logging
 
 import scala.collection.JavaConverters._
 
-class ProcessJob(commandBuilder: CommandBuilder, jobCoordinator: JobModelManager) extends StreamJob with Logging {
-  var jobStatus: Option[ApplicationStatus] = None
-  var process: Process = null
-
-  def submit: StreamJob = {
-    jobStatus = Some(New)
-    val waitForThreadStart = new CountDownLatch(1)
+object ProcessJob {
+  private def createProcessBuilder(commandBuilder: CommandBuilder): ProcessBuilder = {
     val processBuilder = new ProcessBuilder(commandBuilder.buildCommand.split(" ").toList.asJava)
+    processBuilder.environment.putAll(commandBuilder.buildEnvironment)
+
+    // Pipe all output to this process's streams.
+    processBuilder.redirectOutput(ProcessBuilder.Redirect.INHERIT)
+    processBuilder.redirectError(ProcessBuilder.Redirect.INHERIT)
 
     processBuilder
-      .environment
-      .putAll(commandBuilder.buildEnvironment)
+  }
+}
+
+class ProcessJob(commandBuilder: CommandBuilder, val jobModelManager: JobModelManager) extends StreamJob with Logging {
+
+  import ProcessJob._
+
+  val lock = new Object
+  val processBuilder: ProcessBuilder = createProcessBuilder(commandBuilder)
+  var jobStatus: ApplicationStatus = New
+  var processThread: Option[Thread] = None
+
 
-    // create a non-daemon thread to make job runner block until the job finishes.
-    // without this, the proc dies when job runner ends.
-    val procThread = new Thread {
+  def submit: StreamJob = {
+    val threadStartCountDownLatch = new CountDownLatch(1)
+
+    // Create a non-daemon thread to make job runner block until the job finishes.
+    // Without this, the proc dies when job runner ends.
+    processThread = Some(new Thread {
       override def run {
-        process = processBuilder.start
-
-        // pipe all output to this process's streams
-        val outThread = new Thread(new Piper(process.getInputStream, System.out))
-        val errThread = new Thread(new Piper(process.getErrorStream, System.err))
-        outThread.setDaemon(true)
-        errThread.setDaemon(true)
-        outThread.start
-        errThread.start
-        waitForThreadStart.countDown
-        process.waitFor
-        jobCoordinator.stop
+        var processExitCode = -1
+        var process: Option[Process] = None
+
+        setStatus(Running)
+
+        try {
+          threadStartCountDownLatch.countDown
+          process = Some(processBuilder.start)
+          processExitCode = process.get.waitFor
+        } catch {
+          case _: InterruptedException => process foreach { p => p.destroyForcibly }
```

Comment:
Do we need to have a NPE guard here? 

I think process variable will be null If processThread is interrupted before we launch the process.

Reply:
True. The `foreach` on `process` provides this guard. The code in the block is not executed if `process` is `None`. I know it may seem a little counter-intuitive but this was the recommended style on the prescribed [Twitter's Scala coding guidelines](http://twitter.github.io/effectivescala/#Functional%20programming-Options).